dds/DCPS/ConfigStoreImpl.cpp

```diff
@@ -0,0 +1,383 @@
+{
+  config_topic_->connect(config_writer_);
+  config_topic_->connect(config_reader_);
+}
+
+ConfigStoreImpl::~ConfigStoreImpl() {
+  config_topic_->disconnect(config_reader_);
+  config_topic_->disconnect(config_writer_);
+}
+
+CORBA::Boolean
+ConfigStoreImpl::has(const char* key)
+{
+  DCPS::InternalDataReader<ConfigPair>::SampleSequence samples;
+  DCPS::InternalSampleInfoSequence infos;
+  config_reader_->read_instance(samples, infos, DDS::LENGTH_UNLIMITED, ConfigPair(key, ""),
+                                DDS::ANY_SAMPLE_STATE, DDS::ANY_VIEW_STATE, DDS::ALIVE_INSTANCE_STATE);
+  for (size_t idx = 0; idx != samples.size(); ++idx) {
+    const DDS::SampleInfo& info = infos[idx];
+    if (info.valid_data) {
+      return true;
+    }
+  }
+
+  return false;
+}
+
+void
+ConfigStoreImpl::set_boolean(const char* key,
+                             CORBA::Boolean value)
+{
+  set_string(key, value ? "1" : "0");
+}
+
+CORBA::Boolean
+ConfigStoreImpl::get_boolean(const char* key,
+                             CORBA::Boolean value)
+{
+  DCPS::InternalDataReader<ConfigPair>::SampleSequence samples;
+  DCPS::InternalSampleInfoSequence infos;
+  config_reader_->read_instance(samples, infos, DDS::LENGTH_UNLIMITED, ConfigPair(key, ""),
+                                DDS::ANY_SAMPLE_STATE, DDS::ANY_VIEW_STATE, DDS::ALIVE_INSTANCE_STATE);
+  for (size_t idx = 0; idx != samples.size(); ++idx) {
+    const ConfigPair& sample = samples[idx];
+    const DDS::SampleInfo& info = infos[idx];
+    if (info.valid_data) {
+      // We could support "true" and "false".
+      CORBA::Boolean x = 0;
+      if (DCPS::convertToInteger(sample.value(), x)) {
+        value = x;
+      } else if (log_level >= LogLevel::Warning) {
+        ACE_ERROR((LM_WARNING,
+                   ACE_TEXT("(%P|%t) WARNING: ConfigStoreImpl::get_boolean: ")
+                   ACE_TEXT("failed to parse boolean for %C=%C\n"),
+                   sample.key().c_str(), sample.value().c_str()));
+      }
+    }
+  }
+
+  return value;
+}
+
+void
+ConfigStoreImpl::set_int32(const char* key,
+                           CORBA::Long value)
+{
+  set_string(key, to_dds_string(value).c_str());
+}
+
+CORBA::Long
+ConfigStoreImpl::get_int32(const char* key,
+                           CORBA::Long value)
+{
+  DCPS::InternalDataReader<ConfigPair>::SampleSequence samples;
+  DCPS::InternalSampleInfoSequence infos;
+  config_reader_->read_instance(samples, infos, DDS::LENGTH_UNLIMITED, ConfigPair(key, ""),
+                                DDS::ANY_SAMPLE_STATE, DDS::ANY_VIEW_STATE, DDS::ALIVE_INSTANCE_STATE);
+  for (size_t idx = 0; idx != samples.size(); ++idx) {
+    const ConfigPair& sample = samples[idx];
+    const DDS::SampleInfo& info = infos[idx];
+    if (info.valid_data) {
+      CORBA::Long x = 0;
+      if (DCPS::convertToInteger(sample.value(), x)) {
+        value = x;
```

Comment:
`break` after this?

Reply:
This is me being paranoid.  If we ever change the history depth qos, then we could end up with multiple values for one key.  This code ensures that the last (most recent) value is the one returned.